src/DBFTPlugin/ConsensusMessage.cs

```diff
@@ -30,18 +30,29 @@ public virtual void Deserialize(BinaryReader reader)
                 throw new FormatException();
             BlockIndex = reader.ReadUInt32();
             ValidatorIndex = reader.ReadByte();
-            if (ValidatorIndex >= DBFTPlugin.System.Settings.ValidatorsCount)
-                throw new FormatException();
             ViewNumber = reader.ReadByte();
         }
 
-        public static ConsensusMessage DeserializeFrom(byte[] data)
+        public static ConsensusMessage DeserializeFrom(byte[] data, byte validatorsCount)
         {
-            ConsensusMessageType type = (ConsensusMessageType)data[0];
-            Type t = typeof(ConsensusMessage);
-            t = t.Assembly.GetType($"{t.Namespace}.{type}", false);
-            if (t is null) throw new FormatException();
-            return (ConsensusMessage)data.AsSerializable(t);
+            using MemoryStream ms = new(data, false);
+            using BinaryReader reader = new(ms, Utility.StrictUTF8);
+
+            ConsensusMessageType t = (ConsensusMessageType)reader.ReadByte();
+            ConsensusMessage message = t switch
+            {
+                ConsensusMessageType.PrepareRequest => new PrepareRequest(),
+                ConsensusMessageType.PrepareResponse => new PrepareResponse(),
+                ConsensusMessageType.ChangeView => new ChangeView(),
+                ConsensusMessageType.Commit => new Commit(),
+                ConsensusMessageType.RecoveryRequest => new RecoveryRequest(),
+                ConsensusMessageType.RecoveryMessage => new RecoveryMessage(validatorsCount),
+                _ => throw new FormatException(),
+            };
+            message.Deserialize(reader);
```

Comment:
I don't think it works. You have read the `ConsensusMessageType` once before, and you will read it again in `Deserialize()`.

Reply:
Done.